### PR TITLE
DBZ-3364 add additional metadata to fields for API spec generator for the Debezium UI

### PR DESCRIPTION
**Pull request by rk3rn3r**
This is the first part of splitting up PR #2273 into smaller parts, and continue from that PoC PR.

part of https://issues.redhat.com/browse/DBZ-3364

### CODE REVIEW

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 1]
```diff
@@ -115,6 +115,7 @@ public static SnapshotMode parse(String value, String defaultValue) {
     public static final Field HOSTS = Field.create("mongodb.hosts")
             .withDisplayName("Hosts")
             .withType(Type.LIST)
+            .withGroup(Field.createGroupEntry(Field.Group.CONNECTION, 1))
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:118]: So why do we need this in addition to the existing grouping mechanism?

**Review by gunnarmorling — COMMENTED** [review]
@rk3rn3r, one question inline.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java (hunk 1 above)
Comment (rk3rn3r) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:118]: It reflects the groups for the Debezium UI. Later we might use these as new groups.

Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:118]: Yes, my question is: why don't we simply re-group the properties as per what's used for Debezium UI? Do you consider this to be too disruptive at this point?

Comment (rk3rn3r) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:118]: Exactly. I thought we had this discussion before and decided to first introduce it like that and later switch these groups to be used for the Kafka Connect groups.

Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:118]: Ok, that works for me.

**Comment by gunnarmorling** [comment]
@jpechane, can you take a look too. It shouldn't cause problems, but given it touches all the config definitions, a second pair of eye balls is appreciated. Note this change by itself doesn't provide any value yet, it's rather the prerequisite for the API spec generator that will be created in a next step, which will create a JSON Schema definition of all configuration options, which then are exposed from KC to the DBZ UI backend.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 2]
```diff
@@ -124,9 +125,10 @@ public static SnapshotMode parse(String value, String defaultValue) {
     public static final Field LOGICAL_NAME = Field.create("mongodb.name")
             .withDisplayName("Namespace")
             .withType(Type.STRING)
+            .withGroup(Field.createGroupEntry(Field.Group.CONNECTION, 0))
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:128]: The position in group - is it necessary to have it manually assigned? If yes

Are gaps in the allowed?
Are position duplicates allowed?

If an answer to any of the above questions is no then I belive there should be a validation ensuring the rules are not violated.

**Comment by jpechane** [comment]
@gunnarmorling A question related to numbering added.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:128]: > The position in group - is it necessary to have it manually assigned?

@jpechane, see the discussion above: yes, the manual ordering is required for the time being, as we felt it'd be too disruptive at this point to re-group all the fields. It's something we should do eventually though.

> Are gaps in the allowed?

Yes, that shouldn't matter.

> Are position duplicates allowed?

Ideally not. This sequencing is used for ordering the fields in the Debezium UI, so having duplicates would create an ambiguous situation. So +1 for validating that there are no duplicates, good catch.

@rk3rn3r, can you push a commit for ^^^? Thx!

Comment (rk3rn3r) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:128]: Hmmm currently used positions per group are not stored in a list where I can easily lookup collisions. I thought about resolving collisions on read with the API spec generator. Should I really add it to debezium-core code and log all group/positions in a list and look it up? Maybe I can instead write test in core which gets executed for all connectors that checks all fields of a connector for duplicate group-positions? WDYT?

**Comment by gunnarmorling** [comment]
@rk3rn3r, about validating the order, I'm fine with doing this in the generator. In that light, can we merge this one then?

**Comment by rk3rn3r** [comment]
@gunnarmorling I was thinking about this yesterday. I am leaning towards adding unit tests that validate duplicates per connector. so we realize on every PR when a connector has duplicate positions. So we keep the production code slim but have validation (earlier than only when executing the generator).

**Comment by gunnarmorling** [comment]
That's fine for me too. Whatever floats the boat. Let's just try and get this done for next week's release. Thanks :)

**Comment by rk3rn3r** [comment]
Great! I'm close to having a junit test that can give output like that:
```
[ERROR] Failures: 
[ERROR]   FieldTest>AbstractFieldTest.shouldNotAllowDuplicateGroupAssignment:50 Duplicate field group assingnments found: "signal.data.collection" uses an occupied position in group "ADVANCED". Position no. 19 is already taken by field: "source.struct.version
```

File: debezium-core/src/test/java/io/debezium/config/AbstractFieldTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.List;
+import java.util.Map;
+
+import org.junit.Test;
+
+/**
+ * @author Randall Hauch
+ *
+ */
+public abstract class AbstractFieldTest {
```
Comment (rk3rn3r) [inline debezium-core/src/test/java/io/debezium/config/AbstractFieldTest.java:21]: I added this to test for duplicates. It's an abstract implementation here in core that's used by the different connectors to test for duplicates in the respective connector config (ALL_FIELDS).

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FieldTest.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package io.debezium.connector.mongodb;
+
+import org.junit.Before;
+
+import io.debezium.config.AbstractFieldTest;
+
+public class FieldTest extends AbstractFieldTest {
```
Comment (rk3rn3r) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FieldTest.java:7]: As example: this is the concrete instance for duplicate testing for Mongo DB.

**Comment by gunnarmorling** [comment]
@rk3rn3r, could you rebase this PR, addressing the merge conflict? Other than that, is it good to go from your PoV?

**Review by gunnarmorling — APPROVED** [review]

**Comment by gunnarmorling** [comment]
Applied; thx a lot, @rk3rn3r! Looking forward to the follow-ups and the usage for the UI 🚀 .
